lib/index.js

```diff
@@ -43,46 +43,118 @@ class Flooper{
     this.curBlock = 0;
     this.order = this.flooperItemCount + 1;
     this.speed = speed;
+    this.lastKnownSpeed = this.speed;
 
-    // Bindings
-    this.step = this.step.bind(this);
+    /** 
+     * Bindings
+     */
+    // CORE
+    this.init = this.init.bind(this);
     this.checkBlocks = this.checkBlocks.bind(this);
-  }
-
-  // init
-  init(){
-    requestAnimationFrame(this.step);
-  }
-
-  step(timeStamp){
-    this.left -= this.speed;
-    this.checkBlocks();
+    this.step = this.step.bind(this);
+    this.update = this.update.bind(this);
+    // perf / rAF
+    this.requestTick = this.requestTick.bind(this);
+    this.ticking = false;
 
-    this.flooper.style[this.transformProp] = `translateX(${this.left}px)`
+    // ANIMATION / INTERACTION
+    this.play = this.play.bind(this);
+    this.pause = this.pause.bind(this);
+    this.resume = this.resume.bind(this);
+    this.slowMotion = this.slowMotion.bind(this);
 
-    requestAnimationFrame(this.step);
+    // EVENTS
+    this.addEvents = this.addEvents.bind(this);
   }
 
+    // init
+    init(){
+      this.addEvents();
+      this.play()
+    }
+
+  /**
+   * --------------------
+   * CORE
+   * --------------------
+   */
+  // THE MAGIC
   checkBlocks(){
     // cache block
     let block = this.flooperItem[this.curBlock];
     // cache dimensionsObj
-    let blockSize = block.getBoundingClientRect();
+    let blockRect = block.getBoundingClientRect();
+    let blockComputedStyle = window.getComputedStyle(block);
+    // Check if block has margins
+    let blockSpacings = parseInt(blockComputedStyle.marginLeft, 10) + parseInt(blockComputedStyle.marginRight, 10); 
+    let blockComputedWidth = blockRect.width + blockSpacings;
 
-    if(blockSize.x < -(blockSize.width + this.bufferSize) ) {
+    if(blockRect.x < -(blockComputedWidth + this.bufferSize) ) {
       block.style.order=this.order;
       this.order ++;
-      this.left+=blockSize.width;
+      this.left+=blockComputedWidth;
       this.curBlock++;
 
       if(this.curBlock > this.flooperItemCount - 1){
         this.curBlock = 0;
       }
     }
   }
-}
 
 
-// instanciate and initizialize
-var i = new Flooper();
-i.init();
+  /* rAF */
+  update() {
+    // reset the tick so we can
+    // capture the next onScroll
+    this.ticking = false;
+
+    this.checkBlocks();
+    this.flooper.style[this.transformProp] = `translateX(${this.left}px)`;
+
+    requestAnimationFrame(this.step);
+  }
+
+  step() {
+      this.left -= this.speed; //No IE8
+
+      this.requestTick();
+  }
+
+  requestTick() {
+    if(!this.ticking) {
+      requestAnimationFrame(this.update);
+    }
+    this.ticking = true;
+  }
+
+  /**
+   * --------------------
+   * ANIMATION
+   * --------------------
+   */
+  slowMotion(){
+    this.lastKnownSpeed = this.speed;
+    this.speed = this.speed/2;
+  }
+
+  play(){
+    requestAnimationFrame(this.step);
+  }
+
+  pause(){
+    this.lastKnownSpeed = this.speed;
+    this.speed = 0;
```

Comment:
This means the requestAnimationFrame keeps on going, right? Wouldn't it be better to have a boolean wrapping line 114 preventing the next requestAnimationFrame from being called?

Reply:
True my friend, true. I said more or less that in commit 10ced97. I just couldn't make it work, was super tired and it was 3 in morning and i need it for Ably x). If you have time be my guest and update this PR @tomasmcm please!

Related:
- https://stackoverflow.com/questions/30290404/why-is-cancelanimationframe
- http://jsfiddle.net/epistemex/rnqLfz14/29/